presto-main/src/main/java/com/facebook/presto/execution/SqlTask.java

```diff
@@ -139,20 +140,28 @@ public void stateChanged(TaskState newState)
             implements SystemMemoryUsageListener
     {
         private final QueryContext queryContext;
+        private final TaskId taskId;
+        private TaskContext taskContext;
 
-        public UpdateSystemMemory(QueryContext queryContext)
+        public UpdateSystemMemory(QueryContext queryContext, TaskId taskId)
         {
             this.queryContext = requireNonNull(queryContext, "queryContext is null");
+            this.taskId = requireNonNull(taskId, "taskId is null");
         }
 
         @Override
         public void updateSystemMemoryUsage(long deltaMemoryInBytes)
         {
+            if (taskContext == null) {
+                Optional<TaskContext> taskContext = queryContext.getTaskContextByTaskId(taskId);
```

Comment:
Consider adding `queryContext.getRequiredTaskContext(taskId)` to do this in one shot without the `Optional`